Fixed autocommit calls for mysql-connector-python

ericpp opened the pull request:
The MySQLdb and mysql-connector-python clients use different methods for getting and setting the autocommit mode. This code checks the `conn` param passed into get/set_autocommit() to see if it's a MySQLdb instance (containing the get_autocommit method) or a mysql-connector-python instance (missing get_autocommit) and makes the appropriate autocommit calls for the client.

Fixes #14857

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

dstandish commented on airflow/providers/mysql/hooks/mysql.py:
```diff
@@ -52,7 +52,10 @@ def __init__(self, *args, **kwargs) -> None:
 
     def set_autocommit(self, conn: Connection, autocommit: bool) -> None:  # noqa: D403
```
in the code here, presently this `Connection` refers to the airflow model connection but this is not correct

it should be this one: `from MySQLdb.connections import Connection`

now might be a good opportunity to fix this, and also make it a `Union` with the mysql-connector connection type also

ericpp replied:
Do you have to import both modules before you can use their names for a `Union` typing? If so, I would need to figure out if having them both imported at the same time would cause any issues. It looks like the `get_conn()` method is only importing one module depending on the value of `client_name`

dstandish replied:
again this is a good point.

the `mysql` extra installs both connector python and mysqldb so it should be safe to import both.  i don't think we save much by importing locally.

i'll ask around though.  please write a test for this also -- it (your autocommit fix) is testable behavior and i think merits it

ashb replied:
Generally: one of the biggest slow downs in Airflow "boot" time is importing modules, so avoiding top level imports where it's not strictly needed is better I think.

dstandish replied:
Ok, given that, I guess we should just remove that type annotation

dstandish replied:
Ash provided another option in slack ....

You could put the annotation import behind `if TYPE_CHECKING:`

I think either way is ok

dstandish replied:
btw i explored this idea...


```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from mysql.connector.abstracts import MySQLConnectionAbstract
    from MySQLdb.connections import Connection as MySQLdbConnection
...
    def set_autocommit(
        self, conn: Union['MySQLdbConnection', 'MySQLConnectionAbstract'], autocommit: bool
    ) -> None:
```

ericpp replied:
Added in recent commits, but not sure how to actually test it.

dstandish replied:
ok i can try to look later today

dstandish replied:
@ericpp i know you were working through the right way to test this autocommit fix, but it seems you added a bunch of tests that are unrelated to this change?  can you remove the tests that don't have anything to do with this PR? perhaps i'm missing something?

after you do that, i'll take a look and see if we need to make any adjustments, or if you have specific questions i can try to address them.

ericpp replied:
The `TestMySqlHookMySqlConnectorPython` class is a version of the existing `TestMySqlHook` class that works for the `mysql-connector-python` client. It seems like the existing `TestMySqlHook` class only supports the `MySQLdb` client. 

I moved the code that works for both clients from `TestMySqlHook` into `TestMySqlBulk`

I did add a `test_get_autocommit` method to both test classes and can remove that if needed.

dstandish replied:
oh i see thanks lemme look

dstandish replied:
> I moved the code that works for both clients from TestMySqlHook into TestMySqlBulk

ok maybe not 'moved' but 'copied'

ericpp replied:
The test_bulk methods were moved into the `TestMySqlBulk` class. The git diff looks strange because it plopped my changes into middle of the old `TestMySql` class.

ericpp replied:
I just removed the `test_get_autocommit` test that I did add to the code.

dstandish commented on airflow/providers/mysql/hooks/mysql.py:
```diff
@@ -52,7 +52,10 @@ def __init__(self, *args, **kwargs) -> None:
+        if hasattr(conn, 'get_autocommit'):  # mysqlclient
+            conn.autocommit(autocommit)
+        else:  # mysql-connector-python
+            conn.autocommit = autocommit
```
it's a little awkward to check for attr `get_autocommit` and then not use that attr  but rather to only use it as a mechanism to infer that the `conn` variable is mysqldb and therefore `autocommit` must be callable

better would be to just check if autocommit is callable directly:

```suggestion
        if callable(conn.autocommit):
            conn.autocommit(autocommit)
        else:
            conn.autocommit = autocommit
```

ericpp replied:
I avoided doing that because `conn.autocommit` is actually a method with a property decorator around it. It has a side effect of running a `SELECT @@session.autocommit` query whenever you try to access it: https://github.com/mysql/mysql-connector-python/blob/86cceb2bc914362e0990459002976c7e5a271e1d/lib/mysql/connector/connection_cext.py#L148

I also decided against the `callable` function as it seems to not exist in some versions of Python 3. I'm not sure which versions you officially support: https://docs.python.org/3/library/functions.html#callable .

dstandish replied:
All good points lemme take another look

dstandish replied:
This was a tricky one!

I totally agree we should avoid generating the side effect twice.

Here are two options I came up with that I think are a little more direct.

First option is you can check the class of `conn`.  

I am pretty sure that mysql connector will always inherit from this base class:
```python
        if isinstance(conn, MySQLConnectionAbstract):
            conn.autocommit = autocommit
        else:
            conn.autocommit(autocommit)
```
Though alternatively you could check if it's MySQLdb:

```python
from MySQLdb.connections import Connection as MySQLdbConnection
...
if isinstance(conn, MySQLdbConnection):
    conn.autocommit(autocommit)
else:
    conn.autocommit = autocommit
```

And the import alias of MySQLdbConnection can be re-used in resolving that type annotation problem.  I'm not sure what's the best practice re aliasing classes in the event of a collision like we have here but this approach seems reasonable to me.

Option 2 (and I think I like this one) is check if `autocommit` is a property.  You can do so without calling it like so:
```python
        if isinstance(conn.__class__.autocommit, property):
            conn.autocommit = autocommit
        else:
            conn.autocommit(autocommit)
```
I think _also_ checking whether that property has a setter is overkill and certainly not worth the complexity, but in the course of looking into this I learned you can do so with `hasattr(conn.__class__.autocommit, 'fset')`.

But I think I'd go with either of these two approaches because checking for `get_autocommit` and not using it is a little confusing and certainly indirect.

What do you think?

ericpp replied:
Thanks! I like option two because it directly tests the thing it's trying to use.

dstandish commented on airflow/providers/mysql/hooks/mysql.py:
```diff
@@ -63,7 +66,10 @@ def get_autocommit(self, conn: Connection) -> bool:  # noqa: D403
         :rtype: bool
         """
-        return conn.get_autocommit()
+        if hasattr(conn, 'get_autocommit'):  # mysqlclient
```
the comment says `mysqlclient ` but isn't the library actually called `MySQLdb`?  not sure you really need the comments but either way is fine

ericpp replied:
I got those names from the `client_name` variables in the get_conn() method. I agree it makes more sense to use the names of the actual libraries rather than the `client_name`

ericpp replied:
It looks like those are the package names whereas `MySQLdb` and `mysql.connector` are the module names:
https://pypi.org/project/mysqlclient/
https://pypi.org/project/mysql-connector-python/

ericpp replied:
Moved to docblock

dstandish commented on airflow/providers/mysql/hooks/mysql.py:
```diff
@@ -52,7 +52,10 @@ def __init__(self, *args, **kwargs) -> None:
+        else:  # mysql-connector-python
+            conn.autocommit = autocommit
 
     def get_autocommit(self, conn: Connection) -> bool:  # noqa: D403
```
here too i'd suggest fixing this type annotation while you're at it

ericpp replied:
Moved to docblock

ericpp committed:
Fixed autocommit calls for mysql-connector-python

The mysqlclient and mysql-connector-python clients use different methods for getting and setting the autocommit mode.
This code adds checks to get/set_autocommit() to determine if `conn.autocommit` should be used as a property (for mysql-connector-python) or as a method (for mysqlclient)

ericpp committed:
Fixed `conn` type checks in `get_autocommit()` and `set_autocommit()` according to dstandish's suggestion

ericpp committed:
Initial attempt at new autocommit tests

ericpp committed:
Updated docstrings for `get_autocommit` and `set_autocommit` and removed need for `noqa: D403`

dstandish commented on tests/providers/mysql/hooks/test_mysql.py:
```diff
@@ -184,6 +197,20 @@ def setUp(self):
         self.db_hook.get_connection = mock.Mock()
         self.db_hook.get_connection.return_value = self.connection
 
+        class FakeMySQLConnection(object):
```
since airflow is python 3+ you can remove `(object)`

ericpp replied:
I've reworked this to have the fake object extend the `mock.MagicMock` class so that the `mysql.py` property checks pass and so that the mock assertions can be used.

ericpp committed:
Split TestMySqlHook into separate test cases for mysqlclient and mysql-connector-python

uranusjr commented on airflow/providers/mysql/hooks/mysql.py:
```diff
@@ -50,20 +56,36 @@ def __init__(self, *args, **kwargs) -> None:
+        :type bool: True to enable autocommit, False to disable autocommit
+        :rtype: None
+        """
+        if hasattr(conn.__class__, 'autocommit') and isinstance(conn.__class__.autocommit, property):
+            conn.autocommit = autocommit
+        else:
+            conn.autocommit(autocommit)
```
Would this be simpler?

```suggestion
        if callable(conn.autocommit):
            conn.autocommit(autocommit)
        else:
            conn.autocommit = autocommit
```

ericpp replied:
I avoided that because it has the side effect of running a `SELECT @@session.autocommit` query whenever the `autocommit` property is accessed in the `mysql-connector-python` client

See: https://github.com/apache/airflow/pull/14869#discussion_r596598632

ericpp committed:
Reordered import statements

ericpp committed:
Fixed pylint error on False comparison

ericpp committed:
Removed the test_get_autocommit methods I added earlier

dstandish committed:
remove unrelated test changes
